tests/pytorch/test_permutation.py

```diff
@@ -0,0 +1,480 @@
+        num_unpermuted_tokens = sorted_indices.size(0)
+        topk = 1
+    unpermuted_tokens = torch.zeros(
+        [num_unpermuted_tokens, permuted_tokens.shape[-1]],
+        dtype=permuted_tokens.dtype,
+        device=permuted_tokens.device,
+    )
+
+    unpermuted_tokens.index_copy_(0, sorted_indices[: permuted_tokens.size(0)], permuted_tokens)
+    unpermuted_tokens = unpermuted_tokens.reshape(-1, topk, permuted_tokens.size(-1))
+    if probs is not None:
+        unpermuted_tokens = unpermuted_tokens * probs.unsqueeze(-1)
+    unpermuted_tokens = unpermuted_tokens.sum(dim=1)
+    return unpermuted_tokens
+
+
+def dtype_tols(te_dtype: tex.DType) -> Dict[str, float]:
+    """Estimated tolerances for a datatype
+
+    Based on tolerances for torch.testing.assert_close.
+
+    """
+    if te_dtype == tex.DType.kFloat32:
+        return dict(rtol=1.0e-6, atol=1.0e-6)
+    if te_dtype == tex.DType.kFloat16:
+        return dict(rtol=3.0e-3, atol=1.0e-5)
+    if te_dtype == tex.DType.kBFloat16:
+        return dict(rtol=2.0e-2, atol=1.0e-5)
+    if te_dtype == tex.DType.kFloat8E5M2 or te_dtype == tex.DType.kFloat8E4M3:
+        return dict(rtol=2.0e-1, atol=1.0e-1)
+    raise ValueError(f"Unsuppored dtype ({te_dtype})")
+
+
+def fp8_to_fp16(uint8_tensor, e4m3: bool = True):
+    assert uint8_tensor.dtype == torch.uint8, "Input tensor must be uint8"
+
+    float16_tensor = torch.zeros_like(uint8_tensor, dtype=torch.float16)
+
+    sign = (uint8_tensor >> 7) & 1
+    exponent_mask = 0xF if e4m3 else 0x1F
+    if e4m3:
+        exponent = (uint8_tensor >> 3) & exponent_mask
+        mantissa = uint8_tensor & 0x7
+    else:
+        exponent = (uint8_tensor >> 2) & exponent_mask
+        mantissa = uint8_tensor & 0x3
+
+    exponent_bias = 7 if e4m3 else 15
+    mantissa_max = 8.0 if e4m3 else 4.0
+
+    normal_mask = (exponent != 0) & ~(exponent == exponent_mask)
+    actual_exponent = exponent[normal_mask].to(torch.float16) - exponent_bias
+    actual_mantissa = (mantissa[normal_mask].to(torch.float16) + mantissa_max) / mantissa_max
+    float16_tensor[normal_mask] = (
+        ((-1) ** sign[normal_mask].to(torch.float16)) * (2**actual_exponent) * actual_mantissa
+    )
+
+    subnormal_mask = (exponent == 0) & (mantissa != 0)
+    subnormal_exponent = 1 - exponent_bias
+    subnormal_mantissa = mantissa[subnormal_mask].to(torch.float16) / mantissa_max
+    float16_tensor[subnormal_mask] = (
+        ((-1) ** sign[subnormal_mask].to(torch.float16))
+        * (2**subnormal_exponent)
+        * subnormal_mantissa
+    )
+
+    return float16_tensor
+
+
+@pytest.mark.parametrize("te_dtype", _te_dtypes)
+@pytest.mark.parametrize("num_tokens", [4096])
+@pytest.mark.parametrize("num_expert", [8, 16])
+@pytest.mark.parametrize("hidden_size", [4096])
+@pytest.mark.parametrize("topK", [1, 2, 5])
+@pytest.mark.parametrize("num_out_tokens", [None, 4050])
+@pytest.mark.parametrize("with_probs", [True, False])
+def test_permutation(
```

Comment:
Hi, 
I think we should split this test into multiple smaller tests.

Reply:
This test just takes 26.00s on an A100. Do we still need to split it further?
`48 passed, 25 skipped in 26.00s`

Reply:
Done.